src/mesh/p4est_mesh.jl

```diff
@@ -224,3 +350,50 @@ function calc_node_coordinates!(node_coordinates, mapping, trees_per_dimension,
     end
   end
 end
+
+
+# Calculate physical coordinates of each node.
+# Extract corners of each tree, interpolate to requested interpolation nodes,
+# map the resulting coordinates with the specified mapping.
+# 2D version
+function calc_tree_node_coordinates!(node_coordinates::AbstractArray{RealT, 4},
+                                     nodes, mapping,
+                                     vertices, tree_to_vertex) where RealT
+  nodes_in = [-1.0, 1.0]
+  matrix = polynomial_interpolation_matrix(nodes_in, nodes)
+  data_in = Array{RealT, 3}(undef, 2, 2, 2)
+  tmp1 = zeros(RealT, 2, length(nodes), length(nodes_in))
+
+  for tree in 1:size(tree_to_vertex, 2)
+    # Tree vertices are stored in Z-order, ignore z-coordinate in 2D, zero-based indexing
+    data_in[:, 1, 1] .= vertices[1:2, tree_to_vertex[1, tree] + 1]
+    data_in[:, 2, 1] .= vertices[1:2, tree_to_vertex[2, tree] + 1]
+    data_in[:, 1, 2] .= vertices[1:2, tree_to_vertex[3, tree] + 1]
+    data_in[:, 2, 2] .= vertices[1:2, tree_to_vertex[4, tree] + 1]
+
+    multiply_dimensionwise!(
+      view(node_coordinates, :, :, :, tree),
+      matrix, matrix,
+      data_in,
+      tmp1
+    )
+  end
+
+  map_node_coordinates!(node_coordinates, mapping)
+end
+
+function map_node_coordinates!(node_coordinates::AbstractArray{RealT, 4}, mapping) where RealT
```

Comment:
```suggestion
function map_node_coordinates!(node_coordinates::AbstractArray{<:Any, 4}, mapping)
```
It seems you don't use `RealT` in the function body. If that's correct, you shouldn't introduce an unnecessary variable name here.